src/webgpu/shader/validation/parse/shadow_builtins.spec.ts

```diff
@@ -0,0 +1,1002 @@
+    keyword: `tanh`,
+    src: `_ = tanh(2);`,
+  },
+  transpose: {
+    keyword: `transpose`,
+    src: `_ = transpose(mat2x2(1, 2, 3, 4));`,
+  },
+  trunc: {
+    keyword: `trunc`,
+    src: `_ = trunc(2);`,
+  },
+  u32: {
+    keyword: `u32`,
+    src: `_ = u32(1i);`,
+  },
+  unpack2x16snorm: {
+    keyword: `unpack2x16snorm`,
+    src: `_ = unpack2x16snorm(2);`,
+  },
+  unpack2x16unorm: {
+    keyword: `unpack2x16unorm`,
+    src: `_ = unpack2x16unorm(2);`,
+  },
+  unpack2x16float: {
+    keyword: `unpack2x16float`,
+    src: `_ = unpack2x16float(2);`,
+  },
+  unpack4x8snorm: {
+    keyword: `unpack4x8snorm`,
+    src: `_ = unpack4x8snorm(4);`,
+  },
+  unpack4x8unorm: {
+    keyword: `unpack4x8unorm`,
+    src: `_ = unpack4x8unorm(4);`,
+  },
+  unpack4xI8: {
+    keyword: `unpack4xI8`,
+    src: `_ = unpack4xI8(4);`,
+  },
+  unpack4xU8: {
+    keyword: `unpack4xU8`,
+    src: `_ = unpack4xU8(4);`,
+  },
+  vec2_templated: {
+    keyword: `vec2`,
+    src: `_ = vec2<f32>(1, 2);`,
+  },
+  vec2: {
+    keyword: `vec2`,
+    src: `_ = vec2(1, 2);`,
+  },
+  vec3_templated: {
+    keyword: `vec3`,
+    src: `_ = vec3<f32>(1, 2, 3);`,
+  },
+  vec3: {
+    keyword: `vec3`,
+    src: `_ = vec3(1, 2, 3);`,
+  },
+  vec4_templated: {
+    keyword: `vec4`,
+    src: `_ = vec4<f32>(1, 2, 3, 4);`,
+  },
+  vec4: {
+    keyword: `vec4`,
+    src: `_ = vec4(1, 2, 3, 4);`,
+  },
+};
+
+g.test('shadow_hides_builtin')
+  .desc(`Test that shadows hide buitins.`)
+  .params(u =>
+    u
+      .combine('builtin', keysOf(kTests))
+      .combine('inject', ['none', 'function', 'sibling', 'module'])
+      .beginSubcases()
+  )
+  .fn(t => {
+    const data = kTests[t.params.builtin];
+    const value = `let ${data.keyword} = 4;`;
+
+    const module_scope = t.params.inject === 'module' ? `var<private> ${data.keyword} : i32;` : ``;
+    const sibling_func = t.params.inject === 'sibling' ? value : ``;
+    const func = t.params.inject === 'function' ? value : ``;
+
+    const code = `
+struct Data {
+  rt_arr: array<i32>,
+}
+@group(0) @binding(0) var<storage> placeholder: Data;
+
+${module_scope}
+
+fn sibling() {
+  ${sibling_func}
+}
+
+@fragment
+fn main() -> @location(0) vec4f {
+  ${func}
+  ${data.src}
+  return vec4f(1);
+}
+    `;
+
+    const pass = t.params.inject === 'none' || t.params.inject === 'sibling';
+    t.expectCompileResult(pass, code);
+  });
+
+const kFloat16Tests = {
+  f16: {
+    keyword: `f16`,
+    src: `_ = f16(2);`,
+  },
+};
+
+g.test('shadow_hides_builtin_f16')
+  .desc(`Test that shadows hide buitins when shader-f16 is enabled.`)
+  .params(u =>
+    u
+      .combine('builtin', keysOf(kFloat16Tests))
+      .combine('inject', ['none', 'function', 'sibling', 'module'])
+
+      .beginSubcases()
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase({ requiredFeatures: ['shader-f16'] });
+  })
+  .fn(t => {
+    const data = kFloat16Tests[t.params.builtin];
+    const value = `let ${data.keyword} = 4;`;
```

Comment:
nit: It took me a little while to figure out what this meant. I'd recommend calling this `fn_shadow`, here and elsewhere

Reply:
Renamed to `local`. This can be used in function and sibling, so calilng it `fn_shadow` is a bit confusing.